### PR TITLE
Fix-dart-runtime

### PR DESCRIPTION
**Pull request by lohanidamodar**
- Updates dart runtimes to work properly
- Adds new dart 2.16 runtime

### CODE REVIEW

File: runtimes/dart-2.14/.packages [hunk 1]
```diff
@@ -3,25 +3,25 @@
...
-string_scanner:file:///home/ionic/.pub-cache/hosted/pub.dartlang.org/string_scanner-1.1.0/lib/
-term_glyph:file:///home/ionic/.pub-cache/hosted/pub.dartlang.org/term_glyph-1.2.0/lib/
-typed_data:file:///home/ionic/.pub-cache/hosted/pub.dartlang.org/typed_data-1.3.0/lib/
+# Generated by pub on 2022-01-25 11:53:26.176453.
+async:file:///home/dlohani/.pub-cache/hosted/pub.dartlang.org/async-2.8.1/lib/
+charcode:file:///home/dlohani/.pub-cache/hosted/pub.dartlang.org/charcode-1.3.1/lib/
+collection:file:///home/dlohani/.pub-cache/hosted/pub.dartlang.org/collection-1.15.0/lib/
+function_types:file:///home/dlohani/.pub-cache/git/function_types-69be4a9c7980c46447a8f35f21b1185caaf3097b/lib/
+http_parser:file:///home/dlohani/.pub-cache/hosted/pub.dartlang.org/http_parser-4.0.0/lib/
+matcher:file:///home/dlohani/.pub-cache/hosted/pub.dartlang.org/matcher-0.12.11/lib/
```
Comment (eldadfux) [inline runtimes/dart-2.14/.packages:12]: Are we sure we need to commit all this local paths?

Comment (PineappleIOnic) [reply runtimes/dart-2.14/.packages:12]: I'm about to push a fix for this 🙂

File: runtimes/dart-2.12/build.sh [hunk 2]
```diff
@@ -3,34 +3,40 @@
...
+    cp /usr/local/src/pubspec.yaml.fallback /usr/local/src/user_code/pubspec.yaml
 fi
 
 # Move to server directory
 cd /usr/local/src
 
-# Get dependencies
-dart pub get
+# Handle user code package dependency
+./prepare
```
Comment (christyjacob4) [inline runtimes/dart-2.12/build.sh:23]: What exactly is the prepare binary ?

File: runtimes/dart-2.12/pubspec.yaml [hunk 3]
```diff
@@ -7,11 +7,6 @@ environment:
   sdk: '>=2.12.0 <3.0.0'
 
 dependencies:
-  path: ^1.8.0
-  alfred: ^0.1.4+5
   function_types:
-    path: ./function_types/
+    git: https://github.com/lohanidamodar/function_types
```
Comment (christyjacob4) [inline runtimes/dart-2.12/pubspec.yaml:11]: Is this required for the runtime ? I believe we were going to wait with the types and add them in our SDK later.

File: tests/Runtimes/RuntimesTest.php [hunk 4]
```diff
@@ -64,111 +64,111 @@ public function setUp(): void
...
-                'tarname' => 'ruby-3-0.tar.gz',
-            ],
-            'swift-5.5' => [
-                'code' => $functionsDir . '/swift.tar.gz',
-                'entrypoint' => 'index.swift',
-                'timeout' => 15,
-                'runtime' => 'swift-5.5',
-                'tarname' => 'swift-5-5.tar.gz',
-            ]
+            // 'node-14.5' => [
```
Comment (christyjacob4) [inline tests/Runtimes/RuntimesTest.php:67]: Looks like a leftover. Can you uncomment this  ?

File: runtimes/dart-2.14/build.sh [hunk 5]
```diff
@@ -3,34 +3,40 @@
...
+    cp /usr/local/src/pubspec.yaml.fallback /usr/local/src/user_code/pubspec.yaml
 fi
 
 # Move to server directory
 cd /usr/local/src
 
-# Get dependencies
-dart pub get
+# Handle pubspec.yaml
+./prepare
```
Comment (christyjacob4) [inline runtimes/dart-2.14/build.sh:23]: we should use a script here. instead of the binary

File: runtimes/dart-2.12/Dockerfile [hunk 6]
```diff
@@ -15,12 +18,15 @@ COPY . /usr/local/src
 
 RUN ["chmod", "+x", "/usr/local/src/launch.sh"]
 RUN ["chmod", "+x", "/usr/local/src/build.sh"]
+RUN ["chmod", "+x", "/usr/local/src/prepare"]
```
Comment (christyjacob4) [inline runtimes/dart-2.12/Dockerfile:21]: Let's stick to using a script to perform the action that `prepare` is doing. I don't think its a good practice to use non standard binaries that perform custom actions. 

Becomes very difficult in terms of future maintenance.

File: runtimes/dart-2.13/Dockerfile [hunk 7]
```diff
@@ -15,12 +15,15 @@ COPY . /usr/local/src
 
 RUN ["chmod", "+x", "/usr/local/src/launch.sh"]
 RUN ["chmod", "+x", "/usr/local/src/build.sh"]
+RUN ["chmod", "+x", "/usr/local/src/prepare"]
```
Comment (christyjacob4) [inline runtimes/dart-2.13/Dockerfile:18]: we should use a script here. instead of the binary

File: runtimes/dart-2.13/build.sh [hunk 8]
```diff
@@ -3,34 +3,40 @@
...
+    cp /usr/local/src/pubspec.yaml.fallback /usr/local/src/user_code/pubspec.yaml
 fi
 
 # Move to server directory
 cd /usr/local/src
 
-# Get dependencies
-dart pub get
+# Handle pubspec.yaml
+./prepare
```
Comment (christyjacob4) [inline runtimes/dart-2.13/build.sh:23]: we should use a script here. instead of the binary

File: runtimes/dart-2.14/Dockerfile [hunk 9]
```diff
@@ -15,12 +15,15 @@ COPY . /usr/local/src
 
 RUN ["chmod", "+x", "/usr/local/src/launch.sh"]
 RUN ["chmod", "+x", "/usr/local/src/build.sh"]
+RUN ["chmod", "+x", "/usr/local/src/prepare"]
```
Comment (christyjacob4) [inline runtimes/dart-2.14/Dockerfile:18]: we should use a script here. instead of the binary

File: runtimes/dart-2.15/Dockerfile [hunk 10]
```diff
@@ -15,12 +15,15 @@ COPY . /usr/local/src
 
 RUN ["chmod", "+x", "/usr/local/src/launch.sh"]
 RUN ["chmod", "+x", "/usr/local/src/build.sh"]
+RUN ["chmod", "+x", "/usr/local/src/prepare"]
```
Comment (christyjacob4) [inline runtimes/dart-2.15/Dockerfile:18]: we should use a script here. instead of the binary

File: runtimes/dart-2.15/build.sh [hunk 11]
```diff
@@ -3,34 +3,40 @@
...
+    cp /usr/local/src/pubspec.yaml.fallback /usr/local/src/user_code/pubspec.yaml
 fi
 
 # Move to server directory
 cd /usr/local/src
 
-# Get dependencies
-dart pub get
+# Handle pubspec.yaml
+./prepare
```
Comment (christyjacob4) [inline runtimes/dart-2.15/build.sh:23]: we should use a script here. instead of the binary

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
I have some concerns over the usage of a binary file when building the runtimes. It's going to be very challenging for maintenance since you cannot really see what's going on inside the binary. We need to replace it with a simple script.

**Comment by christyjacob4** [comment]
There is also an empty file that was committed
`tests/resources/dart/code/main.dart` 
Please check and remove it if it's not relevant

File: tests/Runtimes/RuntimesTest.php (hunk 4 above)
Comment (lohanidamodar) [reply tests/Runtimes/RuntimesTest.php:67]: @PineappleIOnic I did not change anything here?

**Comment by lohanidamodar** [comment]
> There is also an empty file that was committed `tests/resources/dart/code/main.dart` Please check and remove it if it's not relevant
@christyjacob4  I don't see it?

**Review by christyjacob4 — APPROVED** [review]

**Comment by christyjacob4** [comment]
Screenshot 2022-02-09 at 3 50 01 PM
@lohanidamodar  This one ?

**Comment by lohanidamodar** [comment]
> Screenshot 2022-02-09 at 3 50 01 PM @lohanidamodar This one ?

This is not an empty file, this is a test function for running dart runtime tests

**Comment by lohanidamodar** [comment]
@PineappleIOnic I'm not sure why the tests are failing?

**Comment by lohanidamodar** [comment]
@PineappleIOnic The test failure seems to be related to python, i synced with the `refactor` branch, yet it's not passing. If that's ok, you can merge this PR.

**Pull request closed without merging**